snode/src/main/java/org/apache/rocketmq/snode/client/impl/SubscriptionManagerImpl.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public void removePushSession(RemotingChannel remotingChannel) {
+        Set<MessageQueue> subSet = this.clientSubscriptionTable.get(remotingChannel);
+        log.info("subSet: {}", subSet);
+        if (subSet != null) {
+            for (MessageQueue key : subSet) {
+                Set clientSet = pushTable.get(key);
+                if (clientSet != null) {
+                    log.info("Remove push key:{} remoting:{}", key, remotingChannel.remoteAddress());
+                    clientSet.remove(remotingChannel);
+                    if (clientSet.isEmpty()) {
+                        pushTable.remove(key);
+                    }
+                }
+            }
+        }
+        this.clientSubscriptionTable.remove(remotingChannel);
+    }
+
+    @Override
+    public Set<RemotingChannel> getPushableChannel(MessageQueue messageQueue) {
+        return pushTable.get(messageQueue);
+    }
+
+    private Subscription getSubscription(String groupId, ConsumeType consumeType,
+        MessageModel messageModel, ConsumeFromWhere consumeFromWhere) {
+        Subscription subscription = groupSubscriptionTable.get(groupId);
+        if (subscription == null) {
+            subscription = new Subscription();
+            ConcurrentHashMap subscriptionTable = new ConcurrentHashMap<String, SubscriptionData>();
+            subscription.setSubscriptionTable(subscriptionTable);
+            Subscription prev = groupSubscriptionTable.putIfAbsent(groupId, subscription);
+            subscription = prev != null ? prev : subscription;
+        }
+        subscription.setConsumeFromWhere(consumeFromWhere);
+        subscription.setConsumeType(consumeType);
+        subscription.setMessageModel(messageModel);
+        return subscription;
+    }
+
+    private boolean updateSubscribeTopic(Set<SubscriptionData> subscriptionDataSet, Subscription subscription,
+        String groupId) {
+        boolean updated = false;
+        for (SubscriptionData sub : subscriptionDataSet) {
+            SubscriptionData old = subscription.getSubscriptionTable().get(sub.getTopic());
+            if (old == null) {
+                SubscriptionData prev = subscription.getSubscriptionTable().putIfAbsent(sub.getTopic(), sub);
+                if (prev == null) {
+                    updated = true;
+                    log.info("Subscription changed, add new topic, group: {} {}", groupId, sub.toString());
+                }
+            } else if (sub.getSubVersion() > old.getSubVersion()) {
+                if (subscription.getConsumeType() == ConsumeType.CONSUME_PASSIVELY) {
+                    log.info("Subscription changed, group: {} OLD: {} NEW: {}",
+                        groupId,
+                        old.toString(),
+                        sub.toString()
+                    );
+                }
+                subscription.getSubscriptionTable().put(sub.getTopic(), sub);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **128, 129, and 143**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).